openSearchLane.go

```diff
@@ -0,0 +1,425 @@
+	if err != nil {
+		return
+	}
+	osl.openSearchConnection.client = client
+	osl.openSearchConnection.config = config
+
+	if config.LogThreshold == 0 {
+		osl.openSearchConnection.config.LogThreshold = DefaultLogThreshold
+	}
+	if config.MaxBufferSize == 0 {
+		osl.openSearchConnection.config.MaxBufferSize = DefaultMaxBufferSize
+	}
+	if config.BackoffInterval == 0 {
+		osl.openSearchConnection.config.BackoffInterval = DefaultBackoffInterval
+	}
+	if config.BackOffLimit == 0 {
+		osl.openSearchConnection.config.BackOffLimit = DefaultBackOffLimit
+	}
+
+	go osl.openSearchConnection.processConnection(osl)
+
+	return
+
+}
+
+func (osl *openSearchLane) Derive() Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(osl.Context, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) DeriveWithCancel() (Lane, context.CancelFunc) {
+	childCtx, cancelFn := context.WithCancel(context.WithValue(osl.logLane.Context, parent_lane_id, osl.logLane.LaneId()))
+	ll := deriveLogLane(&osl.logLane, childCtx, osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2, cancelFn
+}
+
+func (osl *openSearchLane) DeriveReplaceContext(ctx context.Context) Lane {
+	ll := deriveLogLane(&osl.logLane, context.WithValue(ctx, parent_lane_id, osl.LaneId()), osl.tees, osl.cr)
+
+	osl2 := openSearchLane{
+		openSearchConnection: osl.openSearchConnection,
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl2.logLane)
+	osl2.logLane.writer = log.New(&osl2, "", 0)
+
+	return &osl2
+}
+
+func (osl *openSearchLane) SetMetadata(key string, val string) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	if osl.metadata == nil {
+		osl.metadata = make(map[string]string)
+		osl.metadata["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	}
+
+	osl.metadata[key] = val
+}
+
+func (osl *openSearchLane) Close() {
+	osl.openSearchConnection.detach()
+}
+
+func (osl *openSearchLane) Write(p []byte) (n int, err error) {
+	osl.openSearchConnection.mu.Lock()
+	defer osl.openSearchConnection.mu.Unlock()
+
+	logEntry := string(p)
+
+	if strings.Contains(logEntry, "\n") {
+		logEntry = strings.ReplaceAll(logEntry, "\n", "")
+	}
+
+	// TODO remove this line
+	fmt.Println(logEntry)
+
+	parentLaneId, ok := osl.Context.Value(parent_lane_id).(string)
+	if !ok {
+		parentLaneId = ""
+	}
+
+	logData := &OpenSearchLogMessage{
+		AppName:      osl.openSearchConnection.config.OpenSearchAppName,
+		ParentLaneId: parentLaneId,
+		JourneyID:    osl.journeyId,
+		LaneID:       osl.LaneId(),
+		LogMessage:   logEntry,
+		Metadata:     osl.metadata,
```

Comment:
This could be a problem - the log message could get queued, and then metadata changed, causing data loss in the original log message. Probably should copy the map here and set the timestamp.

Reply:
```
mapCopy := make(map[string]string, len(osl.metadata))
for k,v := range osl.metadata {
   mapCopy[k] = v
}
logData := {
 ...
 Metadata: mapCopy
}
```